File: bioskills/scrnaseq-scanpy-core-analysis/assets/eval/complete_example_analysis.py
"""
Complete Scanpy scRNA-seq Analysis Example — PBMC 3k Dataset
=============================================================

This script demonstrates a complete end-to-end single-cell RNA-seq analysis
using the PBMC 3k dataset from 10X Genomics, following the scrnaseq-scanpy-core-analysis
workflow exactly.

For dataset info: https://support.10xgenomics.com/single-cell-gene-expression/datasets/1.1.0/pbmc3k

Expected runtime: ~10-20 minutes on a standard laptop (first run downloads data ~50MB)
Expected output:  ~2,700 cells, 8-10 clusters, UMAP + marker gene plots

Usage:
    # From the workflow root directory:
    python assets/eval/complete_example_analysis.py

    # With custom output directory:
    python assets/eval/complete_example_analysis.py --output-dir my_results/
"""

import os
import sys
import argparse

# ── Path setup: add scripts/ to Python path ───────────────────────────────────
EVAL_DIR   = os.path.dirname(os.path.abspath(__file__))
SKILL_DIR  = os.path.join(EVAL_DIR, "..", "..")
SCRIPT_DIR = os.path.join(SKILL_DIR, "scripts")
sys.path.insert(0, SCRIPT_DIR)

# ── Default output directory ──────────────────────────────────────────────────
DEFAULT_OUTPUT = os.path.join(SKILL_DIR, "results", "pbmc3k_example")


def run_complete_analysis(output_dir: str = DEFAULT_OUTPUT):
    """
    Run the complete scanpy workflow on PBMC 3k data.

    Steps:
        1. Load PBMC 3k example data
        2. QC metrics, MAD-based filtering, doublet detection
        3. Normalization, HVG selection, PCA
        4. Neighbor graph, Leiden clustering, UMAP
        5. Marker gene detection, cell type annotation
        6. Export results (AnnData, CSVs, plots)
    """

    os.makedirs(output_dir, exist_ok=True)
    print("=" * 70)
    print("COMPLETE SCANPY ANALYSIS — PBMC 3k")
    print("=" * 70)
    print(f"Output directory: {output_dir}\n")

    # ── STEP 1: Load example data ─────────────────────────────────────────────
    print("\n[Step 1] Loading PBMC 3k example data...")
    print("-" * 50)

    from load_example_data import load_example_data
    adata = load_example_data("pbmc3k")
    # Expected: AnnData object with ~2,700 cells × 32,738 genes
    print(f"  Loaded: {adata.n_obs} cells × {adata.n_vars} genes")

    # ── STEP 2: QC metrics, filtering, doublet detection ─────────────────────
    print("\n[Step 2] Quality control...")
    print("-" * 50)

    from qc_metrics import calculate_qc_metrics, batch_mad_outlier_detection
    from filter_cells import run_scrublet_detection, filter_by_mad_outliers

    # Calculate QC metrics (n_genes, total_counts, pct_counts_mt)
    adata = calculate_qc_metrics(adata, species="human")

    # MAD-based adaptive outlier detection (no hard thresholds)
    adata = batch_mad_outlier_detection(adata, batch_key=None)

    # Doublet detection with Scrublet
    adata = run_scrublet_detection(adata, batch_key=None)

    # Apply filters
    n_before = adata.n_obs
    adata = filter_by_mad_outliers(adata, remove_doublets=True)
    n_after = adata.n_obs
    retention = 100 * n_after / n_before
    print(f"  Retained {n_after}/{n_before} cells ({retention:.1f}%)")
    # Expected: ~2,600-2,700 cells retained (>95%)

    # ── STEP 3: Normalize, HVG, scale, PCA ───────────────────────────────────
    print("\n[Step 3] Normalization, HVG selection, PCA...")
    print("-" * 50)

    from normalize_data import run_standard_normalization
    from find_variable_genes import find_highly_variable_genes
    from scale_and_pca import scale_data, run_pca_analysis, suggest_n_pcs

    # Normalize to 10,000 counts per cell + log1p
    adata = run_standard_normalization(adata, target_sum=1e4)

    # Select 2,000 highly variable genes
    adata = find_highly_variable_genes(adata, n_top_genes=2000)

    # Scale (regress out total_counts and pct_counts_mt)
    adata = scale_data(adata, vars_to_regress=["total_counts", "pct_counts_mt"])

    # PCA (50 components)
    adata = run_pca_analysis(adata, n_pcs=50)

    # Get recommended number of PCs for downstream steps
    n_pcs_recommended = suggest_n_pcs(adata)
    print(f"  Recommended PCs for clustering: {n_pcs_recommended}")
    # Expected: typically 10-20 PCs for PBMC 3k

    # ── STEP 4: Neighbor graph, clustering, UMAP ─────────────────────────────
    print("\n[Step 4] Neighbor graph, Leiden clustering, UMAP...")
    print("-" * 50)

    from cluster_cells import build_neighbor_graph, cluster_leiden_multiple_resolutions
    from run_umap import run_umap_reduction
    from plot_dimreduction import plot_umap_clusters

    # Build neighbor graph using PCA embedding
    adata = build_neighbor_graph(
        adata,
        use_rep="X_pca",
        n_neighbors=10,
        n_pcs=n_pcs_recommended
    )

    # Cluster at multiple resolutions (0.4, 0.6, 0.8, 1.0)
    adata = cluster_leiden_multiple_resolutions(
        adata,
        resolutions=[0.4, 0.6, 0.8, 1.0]
    )

    # UMAP for visualization
    adata = run_umap_reduction(adata)

    # Plot UMAP colored by clusters (resolution 0.8 is default)
    plot_umap_clusters(
        adata,
        cluster_key="leiden_0.8",
        output_dir=os.path.join(output_dir, "umap")
    )
    n_clusters = adata.obs["leiden_0.8"].nunique()
    print(f"  Clusters at resolution 0.8: {n_clusters}")
    # Expected: 8-10 clusters for PBMC 3k

    # ── STEP 5: Marker genes and cell type annotation ─────────────────────────
    print("\n[Step 5] Marker gene detection and cell type annotation...")
    print("-" * 50)

    from find_markers import find_all_cluster_markers
    from annotate_celltypes import annotate_celltypes_from_markers
    from plot_dimreduction import plot_umap_gene_expression

    # Find marker genes for each cluster (Wilcoxon rank-sum test)
    markers = find_all_cluster_markers(
        adata,
        cluster_key="leiden_0.8",
        n_genes=25,
        method="wilcoxon"
    )
    print(f"  Marker genes found for {len(markers['cluster'].unique())} clusters")

    # Annotate cell types using known PBMC marker genes
    # (CD3D=T cells, MS4A1=B cells, LYZ=Monocytes, GNLY=NK cells, etc.)
    adata = annotate_celltypes_from_markers(
        adata,
        markers=markers,
        cluster_key="leiden_0.8",
        species="human",
        tissue="PBMC"
    )

    # Plot UMAP with cell type annotations
    plot_umap_clusters(
        adata,
        cluster_key="cell_type",
        output_dir=os.path.join(output_dir, "umap"),
        title="PBMC 3k — Cell Types"
    )

    # Plot canonical marker genes on UMAP
    canonical_markers = ["CD3D", "MS4A1", "LYZ", "GNLY", "PPBP", "FCER1A"]
    plot_umap_gene_expression(
        adata,
        genes=canonical_markers,
        output_dir=os.path.join(output_dir, "markers")
    )

    # ── STEP 6: Export results ────────────────────────────────────────────────
    print("\n[Step 6] Exporting results...")
    print("-" * 50)

    from export_results import export_anndata_results

    export_anndata_results(
        adata,
        markers=markers,
        output_dir=output_dir,
        prefix="pbmc3k"
    )

    # ── Summary ───────────────────────────────────────────────────────────────
    print("\n" + "=" * 70)
    print("✓ COMPLETE ANALYSIS FINISHED")
    print("=" * 70)
    print(f"\nFinal AnnData: {adata.n_obs} cells × {adata.n_vars} genes")
    print(f"Clusters (res=0.8): {n_clusters}")
    print(f"Cell types annotated: {adata.obs['cell_type'].nunique()}")
    print(f"\nOutput files saved to: {output_dir}/")
    print("  pbmc3k.h5ad              — Full AnnData object")
    print("  pbmc3k_metadata.csv      — Cell metadata (clusters, cell types)")
    print("  pbmc3k_markers.csv       — Marker genes per cluster")
    print("  umap/                    — UMAP plots (clusters + cell types)")
    print("  markers/                 — Marker gene expression UMAPs")

    return adata


# ── Expected results reference ─────────────────────────────────────────────────
EXPECTED_RESULTS = {
    "n_cells_raw":      2700,
    "n_cells_filtered": (2600, 2700),   # range after QC
    "n_clusters_08":    (8, 10),         # at resolution 0.8
    "cell_types": [
        "CD4 T cells",
        "CD8 T cells",
        "B cells",
        "NK cells",
        "CD14+ Monocytes",
        "FCGR3A+ Monocytes",
        "Dendritic cells",
        "Platelets"
    ],
    "canonical_markers": {
        "CD4 T cells":         ["CD3D", "CD4", "IL7R"],
        "CD8 T cells":         ["CD3D", "CD8A", "CD8B"],
        "B cells":             ["MS4A1", "CD79A", "CD79B"],
        "NK cells":            ["GNLY", "NKG7", "GZMB"],
        "CD14+ Monocytes":     ["LYZ", "CD14", "CST3"],
        "FCGR3A+ Monocytes":   ["FCGR3A", "MS4A7"],
        "Dendritic cells":     ["FCER1A", "CST3"],
        "Platelets":           ["PPBP", "PF4"]
    }
}


# ── CLI entry point ────────────────────────────────────────────────────────────
if __name__ == "__main__":
    parser = argparse.ArgumentParser(
        description="Run complete Scanpy scRNA-seq analysis on PBMC 3k dataset."
    )
    parser.add_argument(
        "--output-dir", default=DEFAULT_OUTPUT,
        help=f"Output directory (default: {DEFAULT_OUTPUT})"
    )
    args = parser.parse_args()

    adata = run_complete_analysis(output_dir=args.output_dir)

File: scrnaseq-seurat-core-analysis/assets/eval/complete_example_analysis.R
# ============================================================================
# COMPLETE SEURAT ANALYSIS EXAMPLE - PBMC 3K DATASET
# ============================================================================
#
# This script demonstrates a complete end-to-end scRNA-seq analysis
# using the PBMC 3k dataset from 10X Genomics.
#
# For dataset info, see: assets/eval/datasets/pbmc3k_info.md
#
# Expected runtime: ~15-20 minutes on standard laptop

# Load evaluation helpers (for example data)
source("assets/eval/eval_helpers.R")

# Load all required scripts
source("scripts/setup_and_import.R")
source("scripts/qc_metrics.R")
source("scripts/plot_qc.R")
source("scripts/filter_cells.R")
source("scripts/normalize_data.R")
source("scripts/scale_and_pca.R")
source("scripts/cluster_cells.R")
source("scripts/run_umap.R")
source("scripts/plot_dimreduction.R")
source("scripts/find_markers.R")
source("scripts/annotate_celltypes.R")
source("scripts/export_results.R")

# Setup
setup_seurat_libraries()

# ============================================================================
# STEP 1: Load data
# ============================================================================

# Option A: From SeuratData package (easiest)
seurat_obj <- load_seurat_data("pbmc3k")

# Option B: From downloaded 10X data
# seurat_obj <- import_10x_data("path/to/filtered_gene_bc_matrices/hg19/")

# Option C: From H5 file
# seurat_obj <- import_h5_data("path/to/filtered_feature_bc_matrix.h5")

# ============================================================================
# STEP 2: Quality Control
# ============================================================================

# Calculate QC metrics
seurat_obj <- calculate_qc_metrics(seurat_obj, species = "human")

# Visualize QC metrics
plot_qc_violin(seurat_obj, output_dir = "results/qc")
plot_qc_scatter(seurat_obj, output_dir = "results/qc")
plot_qc_histogram(seurat_obj, output_dir = "results/qc")

# Check metrics before filtering
summary(seurat_obj$nFeature_RNA)
summary(seurat_obj$nCount_RNA)
summary(seurat_obj$percent.mt)

# ============================================================================
# STEP 3: Filter cells
# ============================================================================

# Apply tissue-specific filtering (PBMC)
seurat_obj <- filter_cells_by_qc(
  seurat_obj,
  min_features = 200,
  max_features = 2500,
  max_mt_percent = 5
)

# Expected: ~2,600-2,700 cells retained (>95%)

# ============================================================================
# STEP 4: Normalize data
# ============================================================================

# Use SCTransform (recommended for 10X data)
seurat_obj <- run_sctransform(seurat_obj, vars_to_regress = c("percent.mt"))

# Alternative: LogNormalize workflow
# seurat_obj <- run_lognormalize(seurat_obj)
# source("scripts/find_variable_features.R")
# seurat_obj <- find_hvgs(seurat_obj, n_features = 2000)
# seurat_obj <- scale_data(seurat_obj, vars_to_regress = c("percent.mt"))

# ============================================================================
# STEP 5: Dimensionality Reduction - PCA
# ============================================================================

# Run PCA
seurat_obj <- run_pca_analysis(seurat_obj, n_pcs = 50)

# Visualize PCA results
plot_elbow(seurat_obj, output_dir = "results/pca")
plot_pca_heatmaps(seurat_obj, dims = 1:15, output_dir = "results/pca")
plot_pca_loadings(seurat_obj, dims = 1:4, output_dir = "results/pca")

# Decision: Use 30 PCs for clustering (conservative)

# ============================================================================
# STEP 6: Clustering
# ============================================================================

# Test multiple resolutions
seurat_obj <- cluster_multiple_resolutions(
  seurat_obj,
  dims = 1:30,
  resolutions = c(0.4, 0.6, 0.8, 1.0)
)

# Expected clusters:
# - Resolution 0.4: ~7 clusters
# - Resolution 0.8: ~9 clusters
# - Resolution 1.0: ~11 clusters

# Set resolution 0.8 as default (good balance)
Idents(seurat_obj) <- "RNA_snn_res.0.8"

# ============================================================================
# STEP 7: UMAP
# ============================================================================

# Run UMAP for visualization
seurat_obj <- run_umap_reduction(seurat_obj, dims = 1:30)

# Visualize clusters
plot_umap_clusters(seurat_obj, output_dir = "results/umap")

# Compare different resolutions
plot_clustering_comparison(
  seurat_obj,
  resolutions = c(0.4, 0.6, 0.8, 1.0),
  output_dir = "results/umap"
)

# Plot QC metrics on UMAP (check for technical artifacts)
plot_feature_umap(
  seurat_obj,
  features = c("nFeature_RNA", "nCount_RNA", "percent.mt"),
  output_dir = "results/umap"
)

# ============================================================================
# STEP 8: Find marker genes
# ============================================================================

# Find markers for all clusters
all_markers <- find_all_cluster_markers(
  seurat_obj,
  resolution = 0.8,
  min_pct = 0.25,
  logfc_threshold = 0.25
)

# Expected: ~100-200 markers per cluster

# Export marker tables
export_marker_tables(all_markers, output_dir = "results/markers")

# Visualize top markers
plot_top_markers_heatmap(seurat_obj, all_markers, n_top = 10, output_dir = "results/markers")
plot_markers_dotplot(seurat_obj, all_markers, n_top = 5, output_dir = "results/markers")
plot_markers_violin(seurat_obj, all_markers, n_top = 3, output_dir = "results/markers")

# Examine specific cluster markers
head(all_markers[all_markers$cluster == 0, ], 10)

# ============================================================================
# STEP 9: Cell type annotation
# ============================================================================

# Based on marker genes, assign cell type labels
# See references/marker_gene_database.md for marker lists

annotations <- c(
  "0" = "CD4 T cells",           # CD3D+, IL7R+, CD4+
  "1" = "CD14+ Monocytes",       # CD14+, LYZ+, S100A8+
  "2" = "CD8 T cells",           # CD3D+, CD8A+, CD8B+
  "3" = "B cells",               # MS4A1+, CD79A+, CD79B+
  "4" = "NK cells",              # GNLY+, NKG7+, GZMB+
  "5" = "FCGR3A+ Monocytes",     # FCGR3A+, MS4A7+
  "6" = "Dendritic cells",       # FCER1A+, CST3+
  "7" = "Megakaryocytes"         # PPBP+ (if present)
)

seurat_obj <- annotate_clusters_manual(seurat_obj, annotations, resolution = 0.8)

# Visualize annotations
plot_annotated_umap(seurat_obj, output_dir = "results/annotation")
plot_celltype_proportions(seurat_obj, output_dir = "results/annotation")

# Alternative: Automated annotation with SingleR
# seurat_obj <- annotate_with_singler(seurat_obj, reference = "HPCA")

# ============================================================================
# STEP 10: Visualize specific genes
# ============================================================================

# Key immune markers
immune_markers <- c("CD3D", "CD4", "CD8A", "CD14", "MS4A1", "GNLY")
plot_feature_umap(seurat_obj, features = immune_markers, output_dir = "results/umap")

# T cell markers
tcell_markers <- c("CD3D", "CD4", "CD8A", "IL7R", "CCR7", "CD69")
plot_feature_umap(seurat_obj, features = tcell_markers, output_dir = "results/umap")

# ============================================================================
# STEP 11: Export results
# ============================================================================

# Export everything
export_seurat_results(
  seurat_obj,
  output_dir = "pbmc3k_results",
  resolution = 0.8,
  export_counts = TRUE,
  export_metadata = TRUE,
  export_dimreductions = TRUE
)

# Export summary statistics
export_summary_stats(seurat_obj, output_dir = "pbmc3k_results")

# ============================================================================
# ANALYSIS COMPLETE
# ============================================================================

message("\n========================================")
message("PBMC 3k Analysis Complete!")
message("========================================")
message("Results saved in: pbmc3k_results/")
message("\nKey files:")
message("  - seurat_processed.rds (complete Seurat object)")
message("  - cell_metadata.csv (cell annotations)")
message("  - cluster_markers_all.csv (all marker genes)")
message("  - umap_celltypes.svg/png (annotated UMAP)")
message("\nCell types identified:")
print(table(seurat_obj$celltype))

# Save session info for reproducibility
sink("pbmc3k_results/session_info.txt")
sessionInfo()
sink()

message("\nSession info saved to: pbmc3k_results/session_info.txt")
